compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -579,13 +537,6 @@ impl<'tcx> Validator<'_, 'tcx> {
             // `validate_rvalue` upon access.
             Operand::Constant(c) => {
                 if let Some(def_id) = c.check_static_ptr(self.tcx) {
-                    // Only allow statics (not consts) to refer to other statics.
-                    // FIXME(eddyb) does this matter at all for promotion?
-                    let is_static = matches!(self.const_kind, Some(hir::ConstContext::Static(_)));
-                    if !is_static {
-                        return Err(Unpromotable);
-                    }
-
```

Comment:
Why do we want to promote `&STATIC`? The resulting reference is already `'static`, no? Is this doing something else that I'm not seeing? I would have expected us to go the other way and make statics ineligible for promotion in all contexts for now.

Reply:
This enables `let _val: &'static [&'static u32] = &[&FOO2];` in a `fn`.

> I would have expected us to go the other way and make statics ineligible for promotion in all contexts for now.

I'm all in for restricting promotion but I doubt this one can be taken back -- and then it seems strange to not promote this in `fn`.

Reply:
I think this should remain the same for now. Promoting `&STATIC` will cause us to generate more useless promoted MIR bodies for little benefit.

>  I doubt this one can be taken back.

Perhaps, but I would like some time to consider before committing to this in all contexts. `&STATIC` already works everywhere. `static OTHER: &[&u32] = &[&STATIC];` should work without promotion due to the "top-level scope" rule, so promotion is only helpful in very narrow circumstances.

Reply:
> Promoting &STATIC will cause us to generate more useless promoted MIR bodies for little benefit.

That's an argument against most promotion though...
My point is, there is absolutely no reason to do this promotion only in statics. That just makes no sense. And also I was hoping to make this a PR that doesn't require its own crater run with associated 2-3 weeks of waiting.

> static OTHER: &[&u32] = &[&STATIC]; should work without promotion due to the "top-level scope" rule, so promotion is only helpful in very narrow circumstances.

But `TupleStruct(&[&STATIC])` won't work without this. That might be narrow but not more narrow than many other cases here.

Reply:
But, I guess we can do baby steps for now. I'm just not very patient now that I finally see an endgame for this long-standing problem.^^

Reply:
I'm happy to discuss this change as part of its own PR. If I were consistent I suppose I would have demanded that the `&mut []` change be split out as well, but I'm less concerned about that one since promotion actually does lifetime extension and it fits better into my (pipe) dream of requiring that all promotable expressions be valid patterns.

Reply:
`&mut []` is also widely used, there is no taking it back, so it seems just odd to reject it in `const fn`/`const`/`static`.